test/basic.test.js

```diff
@@ -54,49 +68,64 @@ async function testQos (t, qos) {
 
   var publisher = await helper.startClient()
 
-  publisher._client.publish(msg.topic, msg.payload, msg, helper.noError.bind(this, t))
-
-  var messages = await pMap(subscribers, s => helper.receiveMessage(s, t), pMapOptions)
+  function onMessage (client, topic) {
+    var clientId = client._client.options.clientId
+    if (received[clientId]) {
+      t.fail('Duplicated message received')
+    } else {
+      t.equal(topic, msg.topic, 'Message received from ' + clientId)
+      received[clientId] = true
+      if (Object.keys(received).length === 10) {
+        pMap(subscribers, c => c.end(), pMapOptions)
+          .then(() => publisher.end())
+          .catch(t.error.bind(t))
+      }
+    }
+  }
 
-  for (const m of messages) {
-    t.equal(m.topic, msg.topic, 'Message received')
+  for (const sub of subscribers) {
+    sub.on('message', onMessage.bind(this, sub))
   }
 
-  await pMap(subscribers, c => c.end(), pMapOptions)
-  await publisher.end()
+  await publisher.publish(msg.topic, msg.payload, msg)
 }
 
-test('Subscribed clients receive updates - QoS 1', async function (t) {
-  await testQos(t, 1)
+test('Subscribed clients receive updates - QoS 1', function (t) {
+  testQos(t, 1)
+    .catch(t.error.bind(t))
```

Comment:
why this is not using await anymore?

Reply:
For the same reason I told you before... tests were flacky in previous version